### DISCUSSION TITLE
Performance optimization, how to reduce the time spent on "collectGarbage" in the Vulkan backend?

### DISCUSSION THREAD

**Discussion by zy4944796** [google/filament#7042] [General]
Good day everyone, I am trying to optimize the Vulkan backend to make its performance similar to that of the OpenGL backend. I have added systrace to the engine code and compared the systrace files of the OpenGL backend and Vulkan backend. I found that the main reason why the Vulkan backend is slower than the OpenGL backend is because the Vulkan backend needs to call collectGarbage, which takes 1-2ms. Is there any way to optimize this?

This is a comparison of the execution time of FEngine::execute for testing. Please see the yellow and red lines, which represent the time taken for the Vulkan backend to draw 500 cubes and the OpenGL backend to draw 1000 cubes, respectively. The times are similar.

![Screenshot from 2023-08-05 11-01-15](https://github.com/google/filament/assets/8578753/20509d8a-512b-4db4-92eb-f7da3cdc1792)

Through analysis of the Systrace file, it was found that the collectGarbage function called in the commit of the Vulkan backend takes up to 1-2ms of time, with VulkanDisposer::gc accounting for a significant portion of the time, as shown in the following figure:
![Screenshot from 2023-08-05 11-28-53](https://github.com/google/filament/assets/8578753/f7e53d7f-a681-475c-945b-5179de478344)

Why does VulkanDisposer::gc take so much time? Are there any ways to reduce the time taken for this part? For example, by reusing its resources or moving it to another thread?

**Comment by romainguy** [comment]
@poweifeng

**Comment by poweifeng** [comment]
I've opened an issue #7048 for tracking.  I suspect this might be due to a change I made to fix a crash. 
But thank you for the detailed anaylsis @zy4944796 .  I will update the issue with findings and possible solution.
